osm2streets/src/geometry/algorithm.rs

```diff
@@ -80,224 +76,135 @@ pub fn intersection_polygon(
-        return super::degenerate::degenerate(results, iter.next().unwrap(), iter.next().unwrap());
-    }
-
-    if !trim_roads_for_merging.is_empty() {
+        super::degenerate::degenerate(results, iter.next().unwrap(), iter.next().unwrap())
+    } else if !trim_roads_for_merging.is_empty() {
         pretrimmed_geometry(results, roads, &road_lines)
     } else if let Some(result) =
         super::on_off_ramp::on_off_ramp(results.clone(), roads.clone(), road_lines.clone())
     {
         Ok(result)
     } else {
-        generalized_trim_back(results, roads, &road_lines)
+        trim_to_corners(results, roads, sorted_roads)
     }
 }
 
-fn generalized_trim_back(
+/// Handles intersections with at least 3 roads.
+fn trim_to_corners(
     mut results: Results,
     mut roads: BTreeMap<RoadID, InputRoad>,
-    input_road_lines: &[RoadLine],
+    sorted_road_ids: Vec<RoadID>,
 ) -> Result<Results> {
-    let i = results.intersection_id;
-
-    let mut road_lines: Vec<(RoadID, PolyLine)> = Vec::new();
-    for r in input_road_lines {
-        road_lines.push((r.id, r.fwd_pl.clone()));
-        road_lines.push((r.id, r.back_pl.clone()));
-
-        if false {
-            results.debug.push((
-                format!("{} fwd", r.id),
-                r.fwd_pl.make_polygons(Distance::meters(1.0)),
-            ));
-            results.debug.push((
-                format!("{} back", r.id),
-                r.back_pl.make_polygons(Distance::meters(1.0)),
-            ));
-        }
+    // TODO Take Road instead of InputRoad to avoid this
+    let mut sorted_roads = Vec::new();
+    let mut orig_centers = BTreeMap::new();
+    for id in &sorted_road_ids {
+        let road = &roads[id];
+        sorted_roads.push(road.to_road());
+        orig_centers.insert(*id, road.center_line.clone());
     }
 
-    // Intersect every road's boundary lines with all the other lines. Only side effect here is to
-    // populate new_road_centers.
-    let mut new_road_centers: BTreeMap<RoadID, PolyLine> = BTreeMap::new();
-    // TODO If Results has a BTreeMap too, we could just fill this out as we go
-    for (r1, pl1) in &road_lines {
-        // road_center ends at the intersection.
-        let road_center = if roads[r1].dst_i == i {
-            roads[r1].center_line.clone()
-        } else {
-            roads[r1].center_line.reversed()
-        };
+    // Look at every adjacent pair of edges
+    let mut edges = RoadEdge::calculate(sorted_roads.iter().collect(), results.intersection_id);
+    edges.push(edges[0].clone());
 
-        // Always trim back a minimum amount, if possible.
-        let mut shortest_center = road_center.clone();
+    for pair in edges.windows(2) {
+        let one = &pair[0];
+        let two = &pair[1];
 
-        for (r2, pl2) in &road_lines {
-            if r1 == r2 {
-                continue;
-            }
-
-            // If two roads go between the same intersections, they'll likely hit at the wrong
```

Comment:
The regression is because I'm tossing away this code. I don't understand the problem clearly yet. Will fix up in a followup, along with regressions from the placements work in tempe and northgate tests